engine/src/main/java/org/camunda/bpm/engine/impl/repository/CalledProcessDefinitionImpl.java

```diff
@@ -0,0 +1,156 @@
+/*
+ * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
+ * under one or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information regarding copyright
+ * ownership. Camunda licenses this file to you under the Apache License,
+ * Version 2.0; you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.camunda.bpm.engine.impl.repository;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import org.camunda.bpm.engine.repository.ProcessDefinition;
+import org.camunda.bpm.engine.repository.CalledProcessDefinition;
+
+public class CalledProcessDefinitionImpl implements CalledProcessDefinition {
```

Comment:
Does it make sense to reuse `ProcessDefinitionImpl` by using `extends`? I haven't tried it, I was just wondering if we could avoid adding all the fields of that class here again. 

Another way would be to store the `ProcessDefinition` itself as a field here and then use that as a proxy in the `getId()`, `getKey()`, ... methods. 

Not entirely sure yet which one is better. If the referenced process definition changes with regards to the fields that are used here (not entirely sure this is possible but I think it is), those changes would also be reflected in the `CalledProcessDefinitionImpl` if we use the second option. Which again might be exactly what we want or not.

Reply:
I thought about both things too, let me know if you agree with my arguments 😁.

1. As far as I can see `ProcessDefinitionImpl` does not implement `ProcessDefinition`. Only `ProcessDefinitionEntity` implements that. However, it implements and extends a bunch of other stuff, too. Thus, it has quite a large protocol. Extending it would pull that into  `CalledProcessDefinitionImpl` which I don't think is great, because we don't need any of the stuff that is not in `CalledProcessDefinition` and `ProcessDefinition`.

2. I implemented it that way first, but I was not sure if that would lead to the expected behavior. By nature CalledProcessDefinition is somewhat context and time dependent. 
The values in calledFromActivityIds can change over time, e.g., if the call activities bind against latest and the process is no longer the newest one. However, a CalledProcessDefinition object does not reflect those changes. The user has to repeat the query. 
Consequently, non of the other fields should change over time. I think `ProcessDefinitionEntity` and consequently `ProcessDefinition` objects generally do not change after deployment for most fields, except for `isSuspended` which I think can be updated by `ProcessDefinitionEntity#updateModifiableFieldsFromEntity`. Additionally, it could also change for fields in the future, though that might be unlikely. 
Therefor, I decided to create a dead object and copy all the attributes at the cost of a ton of boilerplate code.

Reply:
Thanks for sharing those details and your thoughts. I agree with all arguments for both approaches 👍
As for option 2, we did the same thing for history events for a very similar reason. I think your arguments regarding the `CalledProcessDefinition` being fixed to a specific time and context are sound, so let's keep it as-is. 👍